### PR TITLE
Motion Planning, Gym Wrapper, Compose Agents, Hab2 Bug fixes

### PR DESCRIPTION
**Pull request by ASzot**
## Motivation and Context

The next version of the Habitat 2.0 release, this PR adds motion planning, an OpenAI Gym wrapper, bug fixes, and the initial set of features necessary to compose agent policies to record demonstrations with motion planning. 

## How Has This Been Tested

- Added new tests to `test/test_rearrange_task.py` 
- Added a non-deterministic test to `test/test_motion_plan.py`

## Types of changes

Habitat Baselines Changes:
- Support rendering images from observation with unequal dimensions. 
- Added benchmark_render.py, a new version of benchmark.py which supports rendering videos and saving trajectories to a file. 
- Added Gym wrapper with config for which parts of the observation to get. 
- Added SPA pipeline from Habitat 2.0 paper. 
- Added point cloud code for Planning in PyBullet. 

HAB Task Changes:
- Added more detail to the collision check utilities return values. 
- Added actions for setting absolute kinematic state and target joint states. 
- Updated interactive play script with text. 
- Added separate IK config for the interactive play script. 
- Refactored measurements code to break down the reward signal and reuse measurements for the reward and success signal. 
- Refactored robo versus robot naming. 

Bug fixes:
- Fix bug with size of the joint observation space. 
- Base transform was not being properly calculated. 

- Docs change / refactoring / dependency upgrade
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ASzot** [comment]
Example of motion planning output: 

Uploading episode=None-ckpt=0-rearrangepick_success=1.00.mp4…

File: habitat/tasks/rearrange/actions.py [hunk 1]
```diff
@@ -126,6 +129,40 @@ def step(self, vel, should_step=True, *args, **kwargs):
...
+        return spaces.Box(shape=(7,), low=0, high=1, dtype=np.float32)
+
+    def step(self, set_pos, should_step=True, *args, **kwargs):
+        # No clipping because the arm is being set to exactly where it needs to
+        # go.
+        self._sim: RearrangeSim
+        self._sim.robot.arm_motor_pos = set_pos
+        if should_step:
+            return self._sim.step(HabitatSimActions.ARM_ABS_POS)
+        else:
```
Comment (Skylion007) [inline habitat/tasks/rearrange/actions.py:145]: Else after return is discouraged. Flake8 should complain.

**Comment by ASzot** [comment]
Example of motion planning output: 



https://user-images.githubusercontent.com/10453381/125521302-784cc650-9710-46ec-a29e-6b6a2474d8db.mp4

File: habitat_baselines/agents/benchmark_render.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                    },
+                    tb_writer=self._writer,
+                )
+
+            count_episodes += 1
+
+        if self._traj_save_path is not None:
+            save_dir = osp.dirname(self._traj_save_path)
+            if not osp.exists(save_dir):
+                os.makedirs(save_dir)
```
Comment (Skylion007) [inline habitat_baselines/agents/benchmark_render.py:135]:
```suggestion
            os.makedirs(save_dir, exist_ok=True)
```

File: habitat_baselines/motion_planning/grasp_generator.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+from habitat_baselines.motion_planning.robot_target import (
+    ObjPlanningData,
+    RobotTarget,
+)
+
+
+class GraspGenerator:
+    def __init__(
+        self,
```
Comment (Skylion007) [inline habitat_baselines/motion_planning/grasp_generator.py:13]: missing types

File: habitat_baselines/motion_planning/mp_sim.py [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+from habitat_baselines.motion_planning.robot_target import ObjPlanningData
+from habitat_sim.physics import CollisionGroups, MotionType
+
+
+class MpSim(ABC):
+    """
+    The abstract simulator interface for the motion planner.
+    """
+
+    def __init__(self, sim):
```
Comment (Skylion007) [inline habitat_baselines/motion_planning/mp_sim.py:28]: types missing

File: habitat_baselines/motion_planning/mp_sim.py [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+        pass
+
+    @abstractmethod
+    def get_obj_info(self, obj_idx) -> ObjPlanningData:
+        """
+        Returns information about an object for the grasp planner
+        """
+
+
+class HabMpSim(MpSim):
```
Comment (Skylion007) [inline habitat_baselines/motion_planning/mp_sim.py:118]: types missing

File: habitat_baselines/motion_planning/mp_sim.py [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+    def get_robot_transform(self):
+        return self.robot_transform
+
+    def get_collisions(self, count_obj_colls, ignore_names, verbose):
+        colls = p.getContactPoints(physicsClientId=self.pc_id)
+
+        def convert_coll(x):
+            linkA, bodyA = p.getBodyInfo(x[1], physicsClientId=self.pc_id)
+            linkB, bodyB = p.getBodyInfo(x[2], physicsClientId=self.pc_id)
+            bodyA = bodyA.decode("ascii")
```
Comment (Skylion007) [inline habitat_baselines/motion_planning/mp_sim.py:368]: Hmm, we may want to revisit the API to prevent the need to decode here. It's kinda off for a Python object and easy to fix in PyBind

File: habitat_baselines/motion_planning/mp_spaces.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        joint_lims = self.get_state_lims(True)
+        self._lower_joint_lims, self._upper_joint_lims = (
+            joint_lims[:, 0],
+            joint_lims[:, 1],
+        )
+        self.num_calls = start_num_calls
+        self._should_render = should_render
+
+    def convert_state(self, x):
+        return np.array([x[i] for i in range(7)])
```
Comment (Skylion007) [inline habitat_baselines/motion_planning/mp_spaces.py:93]: Type here? If x is an array the list comprehension is unnecesary cn be used as the getter into x.

File: habitat_baselines/motion_planning/mp_spaces.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+        def getPathLengthObjWithCostToGo(si):
+            obj = ob.PathLengthOptimizationObjective(si)
+            obj.setCostToGoHeuristic(ob.CostToGoHeuristic(admiss_heuristic))
+            return obj
+
+        pdef.setStartAndGoalStates(js_start, js_end)
+        pdef.setOptimizationObjective(getPathLengthObjWithCostToGo(si))
+
+    def render_start_targ(self, render_dir, subdir, targ_state, suffix="targ"):
```
Comment (Skylion007) [inline habitat_baselines/motion_planning/mp_spaces.py:180]: missing types in the whole file

File: habitat_baselines/motion_planning/projector/core.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
+import math
+
+import torch
```
Comment (Skylion007) [inline habitat_baselines/motion_planning/projector/core.py:3]: Avoid making torch as hard import

File: habitat_baselines/motion_planning/robot_target.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import attr
```
Comment (Skylion007) [inline habitat_baselines/motion_planning/robot_target.py:1]: util file is empty?!

**Review by Skylion007 — CHANGES_REQUESTED** [review]
This API is getting very complicated will definitely need extensive refactoring. Or better integration with Habitat-Sims's pybullet. There are a TON of helper classes here with hard coded conversions/ types/ assumptions of parallel lists etc... 

Marked the most obvious and immediate changes.

**Comment by Skylion007** [comment]
Which file here is the gym wrapper? This is quite a big PR, perhaps the gym wrapper should be split off.

**Comment by ASzot** [comment]
> Which file here is the gym wrapper? This is quite a big PR, perhaps the gym wrapper should be split off.

The Gym wrapper is here: https://github.com/facebookresearch/habitat-lab/blob/hab2_motion_planning/habitat_baselines/utils/gym_adapter.py 

I am okay with splitting off the gym wrapper into another PR, but there are anyways a lot of random changes in this PR.

File: habitat/tasks/rearrange/actions.py [hunk 11]
```diff
@@ -126,6 +129,40 @@ def step(self, vel, should_step=True, *args, **kwargs):
             return None
 
 
+@registry.register_task_action
+class ArmAbsPosAction(SimulatorTaskAction):
```
Comment (mathfac) [inline habitat/tasks/rearrange/actions.py:133]: Doc string needed to say how it's different from other arm actions.

File: habitat/tasks/rearrange/actions.py [hunk 12]
```diff
@@ -126,6 +129,40 @@ def step(self, vel, should_step=True, *args, **kwargs):
...
+    @property
+    def action_space(self):
+        return spaces.Box(shape=(7,), low=0, high=1, dtype=np.float32)
```
Comment (mathfac) [inline habitat/tasks/rearrange/actions.py:136]: Magical number 7 should have reasonable constant name so we can change it easily when needed and distinguish it from other 7s.

File: habitat/tasks/rearrange/actions.py (hunk 1 above)
Comment (mathfac) [reply habitat/tasks/rearrange/actions.py:145]: Do you mean it should be `return None` without else?

File: habitat/tasks/rearrange/actions.py [hunk 13]
```diff
@@ -126,6 +129,40 @@ def step(self, vel, should_step=True, *args, **kwargs):
...
+        self._sim: RearrangeSim
+        self._sim.robot.arm_motor_pos = set_pos
+        if should_step:
+            return self._sim.step(HabitatSimActions.ARM_ABS_POS)
+        else:
+            return None
+
+
+@registry.register_task_action
+class ArmAbsPosKinematicAction(SimulatorTaskAction):
```
Comment (mathfac) [inline habitat/tasks/rearrange/actions.py:150]: Doc string for users unfamiliar with Robotics world.

File: habitat/tasks/rearrange/rearrange_sensors.py [hunk 14]
```diff
@@ -70,6 +69,28 @@ def get_observation(self, observations, episode, *args, **kwargs):
         return pos
 
 
+@registry.register_sensor
+class TargetObjectSensor(MultiObjSensor):
```
Comment (mathfac) [inline habitat/tasks/rearrange/rearrange_sensors.py:73]:
```suggestion
class GoalObjectPositionSensor(MultiObjSensor):
```

File: habitat/tasks/rearrange/rearrange_sensors.py [hunk 15]
```diff
@@ -70,6 +69,28 @@ def get_observation(self, observations, episode, *args, **kwargs):
...
+    """
+    This is the ground truth position
```
Comment (mathfac) [inline habitat/tasks/rearrange/rearrange_sensors.py:75]: Example of the good doc string here:
```
Sensor for PointGoal observations which are used in PointGoal Navigation.

    For the agent in simulator the forward direction is along negative-z.
    In polar coordinate format the angle returned is azimuth to the goal.

    Args:
        sim: reference to the simulator for calculating task observations.
        config: config for the PointGoal sensor. Can contain field for
            GOAL_FORMAT which can be used to specify the format in which
            the pointgoal is specified. Current options for goal format are
            cartesian and polar.

            Also contains a DIMENSIONALITY field which specifes the number
            of dimensions ued to specify the goal, must be in [2, 3]

    Attributes:
        _goal_format: format for specifying the goal which can be done
            in cartesian or polar coordinates.
        _dimensionality: number of dimensions used to specify the goal
```

File: habitat/tasks/rearrange/rearrange_sensors.py [hunk 16]
```diff
@@ -184,9 +205,9 @@ def _get_uuid(self, *args, **kwargs):
     def _get_sensor_type(self, *args, **kwargs):
         return SensorTypes.TENSOR
 
-    def _get_observation_space(self, *args, **kwargs):
+    def _get_observation_space(self, *args, config, **kwargs):
         return spaces.Box(
-            shape=(9,),
+            shape=(config.JOINT_DIM,),
```
Comment (mathfac) [inline habitat/tasks/rearrange/rearrange_sensors.py:210]: Glad to see constants moving into config.

File: habitat/tasks/rearrange/rearrange_sensors.py [hunk 17]
```diff
@@ -329,6 +350,31 @@ def update_metric(self, *args, episode, **kwargs):
         self._metric = {idx: dist for idx, dist in zip(idxs, distances)}
 
 
+@registry.register_measure
+class EndEffectorToRestDistance(Measure):
```
Comment (mathfac) [inline habitat/tasks/rearrange/rearrange_sensors.py:354]: Doc string needed describing what resting is about and why it's needed.

File: habitat/tasks/rearrange/rearrange_sim.py [hunk 18]
```diff
@@ -402,6 +407,22 @@ def step(self, action):
 
         return obs
 
+    def viz_pos(self, pos, viz_id=None, r=0.05):
```
Comment (mathfac) [inline habitat/tasks/rearrange/rearrange_sim.py:410]: Longer name and doc string needed.

**Review by mathfac — COMMENTED** [review]
First part of the feedback.

File: habitat_baselines/motion_planning/mp_spaces.py (hunk 7 above)
Comment (ASzot) [reply habitat_baselines/motion_planning/mp_spaces.py:93]: The type is some OMPL internal state vector. I am not sure the proper way to convert it to numpy array.

**Comment by mathfac** [comment]
Let's move GymWrapper to the other PR if possible. That will be easy to review and refer to later.

File: configs/tasks/rearrangepick_replica_cad.yaml [hunk 19]
```diff
@@ -37,13 +37,14 @@ TASK:
         DIMENSIONALITY: 3
     JOINT_SENSOR:
         TYPE: "JointSensor"
+        JOINT_DIM: 7
     EE_SENSOR:
```
[the remark this answers is not in the document]
Comment (ASzot) [reply configs/tasks/rearrangepick_replica_cad.yaml:41]: Resolved some usages, but this will need a follow up PR because "EE" is short for end-effector in many places throughout the code.

**Comment by ASzot** [comment]
I tried separating the GymWrapper PR from this PR, however, unfortunately this PR also contains fixes for the task which are necessary for the GymWrapper tests to pass. Separating the PRs will take a lot more time and I don’t think is worth the effort. 

I removed all code related to motion planning from perception in the PyBullet simulator wrapper, this will be added in a much later PR since the code will require major refactoring. I also added a new *Fetch Reach task* which was useful for my debugging. This PR also now fixes issues with the observation space and reward that I only found after training policies with RL.

File: habitat/tasks/rearrange/rearrange_task.py [hunk 20]
```diff
@@ -7,25 +7,42 @@
...
+
 class RearrangeTask(NavigationTask):
     """
-    Defines additional logic for valid collisions and gripping.
+    Defines additional logic for valid collisions and gripping shared between
+    all rearrangement tasks.
     """
 
+    def overwrite_sim_config(self, sim_config, episode):
+        return merge_sim_episode_with_object_config(sim_config, episode)
```
Comment (mathfac) [inline habitat/tasks/rearrange/rearrange_task.py:32]: Should we call parent `overwrite_sim_config` as well here?

File: habitat/tasks/rearrange/rearrange_sim.py [hunk 21]
```diff
@@ -402,6 +409,22 @@ def step(self, action):
...
```
Comment (mathfac) [inline habitat/tasks/rearrange/rearrange_sim.py:412]: This looks like util method that should be in Simulator class.

**Review by mathfac — COMMENTED** [review]
Next portion.

File: habitat/tasks/rearrange/actions.py [hunk 22]
```diff
@@ -253,8 +311,9 @@ def apply_ee_constraints(self):
             self.robot_ee_constraints[:, 1],
         )
 
-    def set_desired_ee_pos(self, des_rel_pos):
-        self.ee_targ += np.array(des_rel_pos)
+    def set_desired_ee_pos(self, ee_pos):
```
Comment (Skylion007) [inline habitat/tasks/rearrange/actions.py:314]: Call signature of this? Iterable? Array?

**Review by mathfac — APPROVED** [review]
